File: packages/identity/python/caracalai_identity/jwks.py
# Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
# Caracal, a product of Garudex Labs
#
# JWKS cache with 5-min TTL.

import asyncio
import time
import httpx
from typing import Any

_TTL = 300.0


class JwksCache:
    def __init__(self) -> None:
        self._cache: dict[str, tuple[list[dict[str, Any]], float]] = {}
        self._locks: dict[str, asyncio.Lock] = {}
        self._locks_guard = asyncio.Lock()

    async def _lock_for(self, issuer: str) -> asyncio.Lock:
        async with self._locks_guard:
            lock = self._locks.get(issuer)
            if lock is None:
                lock = asyncio.Lock()
                self._locks[issuer] = lock
            return lock

    async def get_keys(self, issuer: str) -> list[dict[str, Any]]:
        url = issuer.rstrip("/") + "/.well-known/jwks.json"
        entry = self._cache.get(issuer)
        if entry and time.monotonic() - entry[1] < _TTL:
            return entry[0]

        # Per-issuer lock coalesces concurrent fetches: the second caller
        # waits, then reads the freshly-cached entry instead of re-fetching.
        lock = await self._lock_for(issuer)
        async with lock:
            entry = self._cache.get(issuer)
            if entry and time.monotonic() - entry[1] < _TTL:
                return entry[0]

            async with httpx.AsyncClient() as client:
                resp = await client.get(url)
                resp.raise_for_status()
                body = resp.json()

            keys: list[dict[str, Any]] = body.get("keys", [])
            self._cache[issuer] = (keys, time.monotonic())
            return keys

File: packages/identity/go/jwks.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// JWKS fetcher with 5-min in-memory cache.

package identity

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"sync"
	"time"
)

const (
	jwksTTL          = 5 * time.Minute
	jwksFetchTimeout = 10 * time.Second
)

type jwksEntry struct {
	keys      map[string]any
	fetchedAt time.Time
}

var (
	jwksMu     sync.RWMutex
	jwksCache  = map[string]*jwksEntry{}
	jwksClient = &http.Client{Timeout: jwksFetchTimeout}
)

// GetJWKS returns the cached key set for issuer, fetching if missing or stale.
func GetJWKS(issuer string) (map[string]any, error) {
	return GetJWKSContext(context.Background(), issuer)
}

// GetJWKSContext is GetJWKS with caller-supplied cancellation.
func GetJWKSContext(ctx context.Context, issuer string) (map[string]any, error) {
	url := issuer + "/.well-known/jwks.json"

	jwksMu.RLock()
	entry, ok := jwksCache[issuer]
	jwksMu.RUnlock()
	if ok && time.Since(entry.fetchedAt) < jwksTTL {
		return entry.keys, nil
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := jwksClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("jwks fetch %d", resp.StatusCode)
	}

	var body struct {
		Keys []json.RawMessage `json:"keys"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, err
	}

	keys := make(map[string]any, len(body.Keys))
	for _, raw := range body.Keys {
		key, kid, err := parseJWK(raw)
		if err != nil {
			continue
		}
		keys[kid] = key
	}

	jwksMu.Lock()
	jwksCache[issuer] = &jwksEntry{keys: keys, fetchedAt: time.Now()}
	jwksMu.Unlock()
	return keys, nil
}

// ResetJWKSCache clears the in-memory JWKS cache. Intended for tests.
func ResetJWKSCache() {
	jwksMu.Lock()
	jwksCache = map[string]*jwksEntry{}
	jwksMu.Unlock()
}

func parseJWK(raw json.RawMessage) (*ecdsa.PublicKey, string, error) {
	var key struct {
		Kty string `json:"kty"`
		Crv string `json:"crv"`
		Kid string `json:"kid"`
		X   string `json:"x"`
		Y   string `json:"y"`
	}
	if err := json.Unmarshal(raw, &key); err != nil {
		return nil, "", err
	}
	if key.Kty != "EC" || key.Crv != "P-256" || key.Kid == "" {
		return nil, "", fmt.Errorf("unsupported jwk")
	}
	xBytes, err := base64.RawURLEncoding.DecodeString(key.X)
	if err != nil {
		return nil, "", err
	}
	yBytes, err := base64.RawURLEncoding.DecodeString(key.Y)
	if err != nil {
		return nil, "", err
	}
	return &ecdsa.PublicKey{Curve: elliptic.P256(), X: new(big.Int).SetBytes(xBytes), Y: new(big.Int).SetBytes(yBytes)}, key.Kid, nil
}
